### PR TITLE
Add support for content-credential command

### PR DESCRIPTION
**Pull request by swadeley**
Add support for content-credential

 In Sat6.8, only the gpg_key option for content-credential is supported.

see 
[hammer gpg sub-command deprecated in favour of content-credentials · Issue #8018 ](https://github.com/SatelliteQE/robottelo/issues/8018)

This should be merged first: https://github.com/SatelliteQE/robottelo/pull/8025

### CODE REVIEW

**Review by mirekdlugosz — APPROVED** [review]
Changes look sane.
I understand you decided to take multi-step way, of first changing existing class and later introducing new class for contentcredential?

Could you add some test results as demonstration this change does what it's supposed to? No need to run entire `test_gpgkey` file, but maybe of tests you have changed in that file.

**Comment by swadeley** [comment]
Hello

Fist a test of the existing `tests/foreman/cli/test_gpgkey.py` file to make sure I have not broken anything:

robottelo]$ git status
On branch change_gpg_to_content-credentials
nothing to commit, working tree clean
robottelo]$ pytest --show-capture=no  tests/foreman/cli/test_gpgkey.py 
========================== test session starts =============================================
platform linux -- Python 3.7.9, pytest-4.6.3, py-1.9.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo
plugins: xdist-2.1.0, services-1.3.1, mock-1.10.4, forked-1.3.0, cov-2.10.1
collecting ... 2020-10-01 18:33:13 - conftest - DEBUG - Collected 28 test cases
collected 28 items                                                                                               

tests/foreman/cli/test_gpgkey.py ............................                                              [100%]
============================ 28 passed, 3 warnings in 1678.09 seconds ===============

then a test of the new file `tests/foreman/cli/test_content_credentials.py`:

robottelo]$ git status
On branch change_gpg_to_content-credentials
nothing to commit, working tree clean
robottelo]$ pytest --show-capture=no tests/foreman/cli/test_content_credentials.py 
================================= test session starts ====================================
platform linux -- Python 3.7.9, pytest-4.6.3, py-1.9.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo
plugins: xdist-2.1.0, services-1.3.1, mock-1.10.4, forked-1.3.0, cov-2.10.1
collecting ... 2020-10-01 18:32:37 - conftest - DEBUG - Collected 28 test cases
collected 28 items                                                                                               

tests/foreman/cli/test_content_credentials.py ............................                                 [100%]
========================= 28 passed, 3 warnings in 1700.63 seconds ===================

File: robottelo/cli/content_credentials.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    create                        Create a content credential
+    delete                        Destroy a content credential
+    info                          Show a content credential
+    list                          List content credentials
+    update                        Update a content credential
+"""
+from robottelo.cli.base import Base
+
+
+class ContentCredential(Base):
```
Comment (mirekdlugosz) [inline robottelo/cli/content_credentials.py:23]: That's ContentCredential or ContentCredentials (in plural)?

I believe usually we name class after hammer command, and they tend to be singular.

File: robottelo/cli/factory.py [hunk 2]
```diff
@@ -391,6 +392,46 @@ def make_gpg_key(options=None):
...
 
+@cacheable
+def make_content_credential(options=None):
+    """Creates a content credential.
+
+    In Satellite 6.8 only gpg_key is supported.
+
+    :param options: Check options using `hammer content-credentials create --help` on satellite.
+
+    :returns GPGKey object
```
Comment (mirekdlugosz) [inline robottelo/cli/factory.py:403]: Does it? Not "`ContentCredential` object"?

File: tests/foreman/cli/test_content_credentials.py [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+The gpg sub-command was deprecated in favour of content-credentials in
+Satellite 6.8
+
+:Requirement: content_credentials
+
+:CaseAutomation: Automated
+
+:CaseLevel: Component
+
+:CaseComponent: content_credentials
```
Comment (mirekdlugosz) [inline tests/foreman/cli/test_content_credentials.py:12]: There's no such component, sorry.
That should be "ContentCredentials".
And while you're at it, please add it to `testimony.yaml`, under `CaseComponent` key. Values there are supposed to be sorted alphabetically (testimony doesn't care, but it's easier for people when order of items is maintained).

**Review by mirekdlugosz — CHANGES_REQUESTED** [review]
We need to add new component to testimony.yaml, or travis will fail for all PRs made after this was merged.

Other than that, looks good.

File: robottelo/cli/content_credentials.py (hunk 1 above)
Comment (swadeley) [reply robottelo/cli/content_credentials.py:23]: Singular:

    `~]# hammer content-credential -h

    Usage:
        hammer [OPTIONS] SUBCOMMAND [ARG] ...`

File: robottelo/cli/factory.py (hunk 2 above)
Comment (swadeley) [reply robottelo/cli/factory.py:403]: Yes, thanks for spotting that:
`:returns ContentCredential object`

File: tests/foreman/cli/test_content_credentials.py (hunk 3 above)
Comment (swadeley) [reply tests/foreman/cli/test_content_credentials.py:12]: OK, and also :Requirement: content_credentials
 needs to be changed I think.

Comment (swadeley) [reply tests/foreman/cli/test_content_credentials.py]: @mirekdlugosz , OK, changes pushed

Thank you

Comment (mirekdlugosz) [reply tests/foreman/cli/test_content_credentials.py]: Requirement field is easier, as it's not transformed or validated in any way. Any string will suffice. If you pick requirement that doesn't exist in Polarion, Jenkins will create it for you during automation run.
Best to go to Polarion and pick one of requirements we already have. If you can't find any close enough, creating new one is OK.

**Review by mirekdlugosz — APPROVED** [review]
My main concern is resolved, ACK

**Comment by swadeley** [comment]
Hello

I will rebase to get https://github.com/SatelliteQE/robottelo/pull/8035 once https://github.com/SatelliteQE/robottelo/pull/8025 is merged so I can them both.

Thank you

**Comment by swadeley** [comment]
Hello

Rebase made. 
I had to force push (probably because I had to fix typo in git commit message) so new test coming soon.

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no tests/foreman/cli/test_content_credentials.py
========================= test session starts ===================================
platform linux -- Python 3.7.9, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo_2/robottelo
plugins: ibutsu-1.0.34, services-1.3.1, mock-1.10.4, forked-1.3.0, xdist-1.34.0
collecting ... 2020-10-13 12:18:06 - conftest - DEBUG - Collected 28 test cases
collected 28 items                                                                                               

tests/foreman/cli/test_content_credentials.py ............................                                 [100%]

=============== 28 passed, 3 warnings in 1721.17s (0:28:41) ==========================

File: robottelo/cli/content_credentials.py (hunk 1 above)
Comment (mshriver) [reply robottelo/cli/content_credentials.py:23]: Its plural in the docblock on L5. Inconsequential, but inconsistent.

File: robottelo/cli/factory.py [hunk 4]
```diff
@@ -391,6 +392,46 @@ def make_gpg_key(options=None):
...
+        os.chmod(key_filename, 0o700)
+        with open(key_filename, 'w') as gpg_key_file:
+            gpg_key_file.write(gen_alphanumeric(gen_integer(20, 50)))
+    else:
+        # If the key is provided get its local path and remove it from options
+        # to not override the remote path
+        key_filename = options.pop('key')
+
+    args = {
+        'key': '/tmp/{0}'.format(gen_alphanumeric()),
```
Comment (mshriver) [inline robottelo/cli/factory.py:421]: Consider using f-strings to generate this string.

File: robottelo/cli/factory.py [hunk 5]
```diff
@@ -391,6 +392,46 @@ def make_gpg_key(options=None):
...
+
+    :returns ContentCredential object
+    """
+    # Organization ID is a required field.
+    if not options or not options.get('organization-id'):
+        raise CLIFactoryError('Please provide a valid ORG ID.')
+
+    # Create a fake gpg key file if none was provided
+    if not options.get('key'):
+        (_, key_filename) = mkstemp(text=True)
```
Comment (mshriver) [inline robottelo/cli/factory.py:411]: May be a personal preference thing, but the `()` are not necessary when unpacking, `_, key_filename = <expr>` would be valid.

File: robottelo/cli/content_credentials.py (hunk 1 above)
Comment (swadeley) [reply robottelo/cli/content_credentials.py:23]: sorry about that, fixing now. The hammer  command uses singular, so we need that for commands, but there will be one other content-credential, so not sure if module needs to be singular

File: robottelo/cli/factory.py (hunk 4 above)
Comment (swadeley) [reply robottelo/cli/factory.py]: OK, change made

File: robottelo/cli/factory.py (hunk 5 above)
Comment (swadeley) [reply robottelo/cli/factory.py:411]: OK. change made. Thank you for the review.

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no tests/foreman/cli/test_content_credentials.py
==================================== test session starts =======================================
platform linux -- Python 3.7.9, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo_2/robottelo
plugins: ibutsu-1.0.34, services-1.3.1, mock-1.10.4, forked-1.3.0, xdist-1.34.0
collecting ... 2020-10-15 18:19:30 - conftest - DEBUG - Collected 28 test cases
collected 28 items                                                                                               

tests/foreman/cli/test_content_credentials.py ............................                                 [100%]


==================== 28 passed, 3 warnings in 2244.48s (0:37:24) =============================

File: tests/foreman/cli/test_content_credentials.py [hunk 6]
```diff
@@ -0,0 +1,880 @@
...
+    if not content:
+        content = gen_alphanumeric(gen_integer(20, 50))
+    with open(key_filename, "w") as gpg_key_file:
+        gpg_key_file.write(content)
+        return key_filename
+
+    return None
+
+
+class TestGPGKey(CLITestCase):
```
Comment (mshriver) [inline tests/foreman/cli/test_content_credentials.py:67]: Why write these based on unittest, instead of going straight to using pytest?  We'll have to refactor this, in short order, to use pytest.

File: tests/foreman/cli/test_content_credentials.py [hunk 7]
```diff
@@ -0,0 +1,880 @@
...
+        """Create a shared organization for all tests to avoid generating
+        hundreds of organizations
+        """
+        super(TestGPGKey, cls).setUpClass()
+        cls.org = make_org(cached=True)
+
+    # Bug verification
+
+    @tier1
+    def test_verify_redmine_4272(self):
```
Comment (mshriver) [inline tests/foreman/cli/test_content_credentials.py:83]: I don't like this pattern of the test function name reflecting a BZ or issue tracker number.  The test function name should provide some context of what is being tested. The issue that is related to the test should be part of its metadata.

File: tests/foreman/cli/test_content_credentials.py [hunk 8]
```diff
@@ -0,0 +1,880 @@
...
+        content = gen_alphanumeric()
+        gpg_key = create_gpg_key_file(content=content)
+        self.assertIsNotNone(gpg_key, 'GPG Key file must be created')
+        gpg_key = make_content_credential(
+            {'key': gpg_key, 'name': gen_string('alpha'), 'organization-id': self.org['id']}
+        )
+        self.assertEqual(gpg_key['content'], content)
+
+    @tier1
+    def test_positive_get_info_by_name(self):
```
Comment (mshriver) [inline tests/foreman/cli/test_content_credentials.py:102]: I think the cases here could be simplified if you wanted to reduce the duplication of actions.

These initial cases are doing slight variations of the same thing, and appear to have a fair amount of duplication in what's being created and checked for existence. Like this get info test could be done at the same time as any of the `exists` checks that are using various name, key, and organization values though unittest subtest.

**Review by mshriver — CHANGES_REQUESTED** [review]
Overall coverage looks good, but could maybe be simplified/parametrized more effectively.

Requesting changes because I don't see a reason to implement this with unittest when we're actively trying to remove that dependency and use pytest only.

File: tests/foreman/cli/test_content_credentials.py (hunk 6 above)
Comment (swadeley) [reply tests/foreman/cli/test_content_credentials.py:67]: @mshriver I did not write them from scratch, they are clones of the gpg key tests. I wanted to get them approved as  quickly as possible so we have test coverage for 6.8 GA. No extra work is involved in my converting them later to not use unittest.

File: tests/foreman/cli/test_content_credentials.py (hunk 7 above)
Comment (swadeley) [reply tests/foreman/cli/test_content_credentials.py:83]: @mshriver it is extremely odd, agree its not good (it came from gpg key file). Will change it to more suitable name.

File: tests/foreman/cli/test_content_credentials.py (hunk 8 above)
Comment (swadeley) [reply tests/foreman/cli/test_content_credentials.py:102]: @mshriver as I explained above, these are just clones of existing gpg key tests to try and get it done ASAP as the new sub-command was only discovered by myself recently.

File: tests/foreman/cli/test_content_credentials.py (hunk 7 above)
Comment (swadeley) [reply tests/foreman/cli/test_content_credentials.py]: Fixed. 

BTW, this is the redmine issue: https://projects.theforeman.org/issues/4272

**Comment by swadeley** [comment]
For the pytest conversion:

robottelo]$ pytest --show-capture=no tests/foreman/cli/test_content_credentials.py 
============================= test session starts ===============================================
platform linux -- Python 3.8.6, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/sjw/sat6/robottelo, inifile: pytest.ini
plugins: ibutsu-1.1.1, mock-3.3.1, forked-1.3.0, services-2.2.1, cov-2.10.1, xdist-1.34.0
collecting ... 2020-11-06 14:40:38 - conftest - DEBUG - Collected 88 test cases
collected 88 items                                                                                               

tests/foreman/cli/test_content_credentials.py ............................................................ [ 68%]
............................                                                                               [100%]



======================== 88 passed, 3 warnings in 2006.16s (0:33:26) ===============================

File: robottelo/cli/base.py [hunk 9]
```diff
@@ -61,67 +61,11 @@ class CLIDataBaseError(CLIBaseError):
...
-        subscription                  Manipulate subscriptions.
-        sync-plan                     Manipulate sync plans
-        task                          Tasks related actions.
-        template                      Manipulate provisioning templates.
-        user                          Manipulate users.
-        user-group                    Manage user groups.
-
-
-    @since: 27.Nov.2013
+    See Subcommands section in `hammer --help` output on your Satellite.
```
Comment (lhellebr) [inline robottelo/cli/base.py:64]: Is this somehow related to the content-credential changes?

**Review by lhellebr — APPROVED** [review]
ACK, nonblocking comment

File: robottelo/cli/base.py (hunk 9 above)
Comment (swadeley) [reply robottelo/cli/base.py:64]: @lhellebr No, but it was very out of date, since 2013, and it seems pointless having out-of-date reference material in the test files.

**Comment by lhellebr** [comment]
Looks good, I'm happy. But Travis is not :(

**Comment by swadeley** [comment]
> Looks good, I'm happy. But Travis is not :(

strangely, when I run Black locally against the file, it changes some lines but does not re-order the imports (which is what Travis complained about). I have also done a rebase, lets see what Travis says.

**Review by mshriver — APPROVED** [review]
Changes look good, gotta fix travis/pre-commit and I'll merge.

**Comment by swadeley** [comment]
Changes just in, thank you for the review

**Comment by swadeley** [comment]
Thank you all
